### PR TITLE
Complete user stories 10, 11, and 12

### PR DESCRIPTION
**Pull request by cmpprg**
- Create user registration form.
- Add sad path coverage for incomplete form.
- Add sad path coverage for same email.
- Make/run migration for user resource.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class User < ApplicationRecord
+  validates :name, presence:true
+  validates :address, presence:true
+  validates :city, presence:true
+  validates :state, presence:true
+  validates :zip, presence:true
+  validates :email, presence:true, uniqueness:true
+  validates :password, presence:true
```
Comment (tony-gomes) [inline app/models/user.rb:8]: I think there might supposed to be a `require: true` here... might be worth inquiring as to its purpose...

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  validates :password_confirmation, presence:true
+
```
Comment (tony-gomes) [inline app/models/user.rb:10]: I am also curious if this password confirmation is a value that is supposed to be validated... seems redundant but also worth inquiring about...

File: config/routes.rb [hunk 3]
```diff
@@ -1,36 +1,42 @@
...
+  delete '/cart/:item_id', to: 'cart#remove_item'
+
+  get '/orders/new', to: 'orders#new'
+  post '/orders', to: 'orders#create'
+  get '/orders/:id', to: 'orders#show'
+  #register
+  get '/register', to: 'register#new'
+  post '/register', to: 'register#create'
+  #profile
+  get '/profile', to: 'profile#show'
```
Comment (tony-gomes) [inline config/routes.rb:41]: Sorry if I missed this, but I was wondering if this should be user_profile?

File: db/migrate/20200406213624_create_user.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  def change
+    create_table :users do |t|
+      t.string :name
+      t.string :address
+      t.string :city
+      t.string :state
+      t.string :zip
+      t.string :email
+      t.string :password_digest
+    end
```
Comment (tony-gomes) [inline db/migrate/20200406213624_create_user.rb:11]: I am still unclear whether we need to explicitly add `t.timestamps` but Users definitely should have them... Can anyone confirm if this is a necessity or if rails adds them by default?

File: db/schema.rb [hunk 5]
```diff
@@ -67,6 +67,16 @@
...
 
+  create_table "users", force: :cascade do |t|
+    t.string "name"
+    t.string "address"
+    t.string "city"
+    t.string "state"
+    t.string "zip"
+    t.string "email"
+    t.string "password_digest"
+  end
```
Comment (tony-gomes) [inline db/schema.rb:78]: I think it is worth getting more clarification about the standard for where User data lives & how it is typically structured.

File: db/migrate/20200406213624_create_user.rb (hunk 4 above)
Comment (tylerpporter) [reply db/migrate/20200406213624_create_user.rb:11]: @tony-gomes I think that is a group preference thing. I don't see timestamps being used in any of the user stories, so I think it's up to us to decide if they would be useful to store in our database.

File: spec/features/users/new_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      fill_in :state, with: "CO"
+      fill_in :zip, with: "80232"
+      fill_in :email, with: "campryan@comcast.net"
+      fill_in :password, with: "password"
+      fill_in :password_confirmation, with: "password"
+      click_button("Submit")
+    end
+
+    expect(current_path).to eql("/register")
+    expect(page).to have_content("Email has already been taken")
```
Comment (tony-gomes) [inline spec/features/users/new_spec.rb:70]: On the refactor we can add the syntax for targeting flash message specifically... I think it is something like this.
`expect(flash[:error]).to be_present`
`expect(flash[:error).to match(/Email has already been taken .*/)`

File: spec/models/merchant_spec.rb [hunk 7]
```diff
@@ -48,7 +48,8 @@
       order_2.item_orders.create!(item: chain, price: chain.price, quantity: 2)
       order_3.item_orders.create!(item: @tire, price: @tire.price, quantity: 2)
 
-      expect(@meg.distinct_cities).to eq(["Denver","Hershey"])
+      expect(@meg.distinct_cities).to include("Hershey")
+      expect(@meg.distinct_cities).to include("Denver")
     end
```
Comment (tony-gomes) [inline spec/models/merchant_spec.rb:53]: https://makandracards.com/makandra/899-check-if-two-arrays-contain-the-same-elements-in-ruby-rspec-or-test-unit

**Review by tony-gomes — APPROVED** [review]
Great first session fellas!

File: app/controllers/register_controller.rb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    if user.save
+      flash[:notice] = "Welcome #{params[:name]} Registration Successful"
+      session[:user_id] = user.id
+      redirect_to "/profile"
+    else
+      flash[:error] = user.errors.full_messages.to_sentence
+      @user_info = user_params
+      render :new
+    end
+  end
```
Comment (tylerpporter) [inline app/controllers/register_controller.rb]: We might want to pull out some of this logic into private helper methods to adhere to Sandi Metz' Rules For Developers.

**Review by tylerpporter — APPROVED** [review]

File: app/models/user.rb (hunk 1 above)
Comment (cmpprg) [reply app/models/user.rb:8]: we received and error that 'RequireValidator' was not a thing. we did try though.

File: config/routes.rb (hunk 3 above)
Comment (cmpprg) [reply config/routes.rb:41]: it calls to /profile explicitly in the user story, so we just went with it.

File: db/schema.rb (hunk 5 above)
Comment (cmpprg) [reply db/schema.rb:78]: agreed.

**Review by tylerpporter — APPROVED** [review]

**Comment by tony-gomes** [comment]
Merging to get the repo to a workable version for our paired kickoff

File: app/controllers/register_controller.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
+class RegisterController < ApplicationController
+  
+  def new
+    @user_info = {}
+  end
+
+  def create
+    user = User.create(user_params)
+    user.save ? success(user) : failure(user)
```
Comment (Shock253) [inline app/controllers/register_controller.rb:9]: interesting refactor of the creation happy path/sad path, I'll have to use this

File: app/controllers/register_controller.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  def success(user)
+    flash[:notice] = "Welcome #{params[:name]} Registration Successful"
+    session[:user_id] = user.id
+    redirect_to "/profile"
+  end
+
+  def failure(user)
+    flash[:error] = user.errors.full_messages.to_sentence
+    @user_info = user_params
+    render :new
```
Comment (Shock253) [inline app/controllers/register_controller.rb:27]: I'm still unsure as to which way is best to do this, both ways seem a little strange

File: db/migrate/20200406213624_create_user.rb (hunk 4 above)
Comment (Shock253) [reply db/migrate/20200406213624_create_user.rb:11]: From what I understand, timestamps are something that we've always put in but haven't used so far, so I would expect it's something we should always put in

**Review by Shock253 — COMMENTED** [review]
Looks good, as long as we ask the instructors what the best practice is for the @user_info section

**Pull request merged**